File: ios/CBVVPN/SOCKS5ProxyHandler.swift
import Foundation
import Network

class SOCKS5ProxyHandler {
    private let proxyHost: String
    private let proxyPort: Int
    private let username: String?
    private let password: String?

    init(host: String, port: Int, username: String?, password: String?) {
        self.proxyHost = host
        self.proxyPort = port
        self.username = username
        self.password = password
    }

    // MARK: - SOCKS5 Protocol Constants

    private enum SOCKS5Constants {
        static let version: UInt8 = 0x05
        static let noAuth: UInt8 = 0x00
        static let usernamePassword: UInt8 = 0x02
        static let connectCommand: UInt8 = 0x01
        static let ipv4AddressType: UInt8 = 0x01
        static let domainAddressType: UInt8 = 0x03
        static let ipv6AddressType: UInt8 = 0x04
        static let success: UInt8 = 0x00
    }

    // MARK: - Connection

    func connect(to destinationHost: String, port: UInt16, completion: @escaping (NWConnection?, Error?) -> Void) {
        // Create connection to proxy server
        let endpoint = NWEndpoint.hostPort(
            host: NWEndpoint.Host(proxyHost),
            port: NWEndpoint.Port(integerLiteral: UInt16(proxyPort))
        )

        let parameters = NWParameters.tcp
        let connection = NWConnection(to: endpoint, using: parameters)

        connection.stateUpdateHandler = { state in
            switch state {
            case .ready:
                self.performSOCKS5Handshake(connection: connection, destinationHost: destinationHost, port: port, completion: completion)
            case .failed(let error):
                completion(nil, error)
            case .cancelled:
                completion(nil, NSError(domain: "SOCKS5", code: -1, userInfo: [NSLocalizedDescriptionKey: "Connection cancelled"]))
            default:
                break
            }
        }

        connection.start(queue: .global(qos: .userInitiated))
    }

    // MARK: - SOCKS5 Handshake

    private func performSOCKS5Handshake(connection: NWConnection, destinationHost: String, port: UInt16, completion: @escaping (NWConnection?, Error?) -> Void) {
        // Step 1: Send authentication methods
        sendAuthenticationMethods(connection: connection) { success, error in
            if let error = error {
                completion(nil, error)
                return
            }

            // Step 2: Receive authentication method selection
            self.receiveAuthMethodSelection(connection: connection) { selectedMethod, error in
                if let error = error {
                    completion(nil, error)
                    return
                }

                // Step 3: Perform authentication if required
                if selectedMethod == SOCKS5Constants.usernamePassword {
                    self.performUsernamePasswordAuth(connection: connection) { success, error in
                        if let error = error {
                            completion(nil, error)
                            return
                        }

                        // Step 4: Send connection request
                        self.sendConnectionRequest(connection: connection, destinationHost: destinationHost, port: port) { success, error in
                            if let error = error {
                                completion(nil, error)
                                return
                            }
                            completion(connection, nil)
                        }
                    }
                } else if selectedMethod == SOCKS5Constants.noAuth {
                    // Step 4: Send connection request (no auth)
                    self.sendConnectionRequest(connection: connection, destinationHost: destinationHost, port: port) { success, error in
                        if let error = error {
                            completion(nil, error)
                            return
                        }
                        completion(connection, nil)
                    }
                } else {
                    completion(nil, NSError(domain: "SOCKS5", code: -1, userInfo: [NSLocalizedDescriptionKey: "Unsupported authentication method"]))
                }
            }
        }
    }

    private func sendAuthenticationMethods(connection: NWConnection, completion: @escaping (Bool, Error?) -> Void) {
        var data = Data()
        data.append(SOCKS5Constants.version) // SOCKS version

        // Add authentication methods
        if username != nil && password != nil {
            data.append(0x02) // Number of methods
            data.append(SOCKS5Constants.noAuth)
            data.append(SOCKS5Constants.usernamePassword)
        } else {
            data.append(0x01) // Number of methods
            data.append(SOCKS5Constants.noAuth)
        }

        connection.send(content: data, completion: .contentProcessed { error in
            if let error = error {
                completion(false, error)
            } else {
                completion(true, nil)
            }
        })
    }

    private func receiveAuthMethodSelection(connection: NWConnection, completion: @escaping (UInt8?, Error?) -> Void) {
        connection.receive(minimumIncompleteLength: 2, maximumLength: 2) { data, _, _, error in
            if let error = error {
                completion(nil, error)
                return
            }

            guard let data = data, data.count == 2 else {
                completion(nil, NSError(domain: "SOCKS5", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid response"]))
                return
            }

            let version = data[0]
            let method = data[1]

            if version != SOCKS5Constants.version {
                completion(nil, NSError(domain: "SOCKS5", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid SOCKS version"]))
                return
            }

            if method == 0xFF {
                completion(nil, NSError(domain: "SOCKS5", code: -1, userInfo: [NSLocalizedDescriptionKey: "No acceptable authentication method"]))
                return
            }

            completion(method, nil)
        }
    }

    private func performUsernamePasswordAuth(connection: NWConnection, completion: @escaping (Bool, Error?) -> Void) {
        guard let username = username, let password = password else {
            completion(false, NSError(domain: "SOCKS5", code: -1, userInfo: [NSLocalizedDescriptionKey: "Username/password not provided"]))
            return
        }

        var data = Data()
        data.append(0x01) // Auth version
        data.append(UInt8(username.count))
        data.append(contentsOf: username.utf8)
        data.append(UInt8(password.count))
        data.append(contentsOf: password.utf8)

        connection.send(content: data, completion: .contentProcessed { error in
            if let error = error {
                completion(false, error)
                return
            }

            // Receive auth response
            connection.receive(minimumIncompleteLength: 2, maximumLength: 2) { data, _, _, error in
                if let error = error {
                    completion(false, error)
                    return
                }

                guard let data = data, data.count == 2 else {
                    completion(false, NSError(domain: "SOCKS5", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid auth response"]))
                    return
                }

                let status = data[1]
                if status == SOCKS5Constants.success {
                    completion(true, nil)
                } else {
                    completion(false, NSError(domain: "SOCKS5", code: -1, userInfo: [NSLocalizedDescriptionKey: "Authentication failed"]))
                }
            }
        })
    }

    private func sendConnectionRequest(connection: NWConnection, destinationHost: String, port: UInt16, completion: @escaping (Bool, Error?) -> Void) {
        var data = Data()
        data.append(SOCKS5Constants.version) // SOCKS version
        data.append(SOCKS5Constants.connectCommand) // CONNECT command
        data.append(0x00) // Reserved

        // Add destination address
        if let ipv4 = IPv4Address(destinationHost) {
            data.append(SOCKS5Constants.ipv4AddressType)
            data.append(contentsOf: ipv4.rawValue)
        } else if let ipv6 = IPv6Address(destinationHost) {
            data.append(SOCKS5Constants.ipv6AddressType)
            data.append(contentsOf: ipv6.rawValue)
        } else {
            // Domain name
            data.append(SOCKS5Constants.domainAddressType)
            data.append(UInt8(destinationHost.count))
            data.append(contentsOf: destinationHost.utf8)
        }

        // Add port
        data.append(UInt8(port >> 8))
        data.append(UInt8(port & 0xFF))

        connection.send(content: data, completion: .contentProcessed { error in
            if let error = error {
                completion(false, error)
                return
            }

            // Receive connection response
            connection.receive(minimumIncompleteLength: 4, maximumLength: 256) { data, _, _, error in
                if let error = error {
                    completion(false, error)
                    return
                }

                guard let data = data, data.count >= 4 else {
                    completion(false, NSError(domain: "SOCKS5", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid connection response"]))
                    return
                }

                let version = data[0]
                let status = data[1]

                if version != SOCKS5Constants.version {
                    completion(false, NSError(domain: "SOCKS5", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid SOCKS version in response"]))
                    return
                }

                if status != SOCKS5Constants.success {
                    let errorMessage = self.getSOCKS5ErrorMessage(status)
                    completion(false, NSError(domain: "SOCKS5", code: Int(status), userInfo: [NSLocalizedDescriptionKey: errorMessage]))
                    return
                }

                completion(true, nil)
            }
        })
    }

    private func getSOCKS5ErrorMessage(_ status: UInt8) -> String {
        switch status {
        case 0x01:
            return "General SOCKS server failure"
        case 0x02:
            return "Connection not allowed by ruleset"
        case 0x03:
            return "Network unreachable"
        case 0x04:
            return "Host unreachable"
        case 0x05:
            return "Connection refused"
        case 0x06:
            return "TTL expired"
        case 0x07:
            return "Command not supported"
        case 0x08:
            return "Address type not supported"
        default:
            return "Unknown SOCKS5 error: \(status)"
        }
    }
}

File: native-templates/android/SOCKS5ProxyHandler.kt
package com.cbv.vpn

import android.util.Log
import java.io.InputStream
import java.io.OutputStream
import java.net.InetSocketAddress
import java.net.Socket

/**
 * Handle SOCKS5 proxy connections
 */
class SOCKS5ProxyHandler(
    val proxyHost: String,
    val proxyPort: Int,
    private val username: String?,
    private val password: String?
) : ProxyHandler {
    
    private val TAG = "SOCKS5ProxyHandler"
    
    override fun connect(targetHost: String, targetPort: Int, socket: Socket): Boolean {
        try {
            Log.d(TAG, "Connecting to SOCKS5 proxy $proxyHost:$proxyPort for $targetHost:$targetPort")
            
            // Check if socket is already connected (e.g., when reusing a socket)
            if (!socket.isConnected) {
                // Configure socket for better performance
                socket.soTimeout = 15_000
                socket.tcpNoDelay = true
                socket.keepAlive = true
                socket.receiveBufferSize = 65536
                socket.sendBufferSize = 65536
                
                Log.d(TAG, "🔌 Connecting socket to proxy...")
                socket.connect(InetSocketAddress(proxyHost, proxyPort), 15_000)
                
                if (!socket.isConnected) {
                    Log.e(TAG, "Failed to connect to proxy")
                    return false
                }
                Log.d(TAG, "✅ Socket connected to proxy")
            } else {
                Log.d(TAG, "✅ Socket already connected to proxy, reusing connection")
            }
            
            val input = socket.getInputStream()
            val output = socket.getOutputStream()
            
            // Step 1: Authentication method negotiation
            if (!username.isNullOrEmpty() && !password.isNullOrEmpty()) {
                // Request username/password auth (support both no-auth and user/pass)
                output.write(byteArrayOf(0x05, 0x02, 0x00, 0x02))
                Log.d(TAG, "Requesting auth methods: NO_AUTH(0x00) and USERNAME_PASSWORD(0x02)")
            } else {
                // Request no auth
                output.write(byteArrayOf(0x05, 0x01, 0x00))
                Log.d(TAG, "Requesting auth method: NO_AUTH(0x00)")
            }
            output.flush()
            
            // Read auth method response
            val authResponse = ByteArray(2)
            if (input.read(authResponse) != 2) {
                Log.e(TAG, "Failed to read auth method response")
                return false
            }
            
            if (authResponse[0] != 0x05.toByte()) {
                Log.e(TAG, "Invalid SOCKS version: ${authResponse[0]}")
                return false
            }
            
            // Step 2: Handle authentication
            when (authResponse[1].toInt() and 0xFF) {
                0x00 -> {
                    Log.d(TAG, "No authentication required")
                }
                0x02 -> {
                    if (username.isNullOrEmpty() || password.isNullOrEmpty()) {
                        Log.e(TAG, "Proxy requires auth but no credentials provided")
                        return false
                    }
                    
                    Log.d(TAG, "Performing username/password authentication")
                    
                    // Send auth request
                    val authRequest = mutableListOf<Byte>()
                    authRequest.add(0x01) // Auth version
                    authRequest.add(username.length.toByte())
                    authRequest.addAll(username.toByteArray().toList())
                    authRequest.add(password.length.toByte())
                    authRequest.addAll(password.toByteArray().toList())
                    
                    output.write(authRequest.toByteArray())
                    output.flush()
                    
                    // Read auth response
                    val authResult = ByteArray(2)
                    if (input.read(authResult) != 2) {
                        Log.e(TAG, "Failed to read auth response")
                        return false
                    }
                    
                    if (authResult[1] != 0x00.toByte()) {
                        Log.e(TAG, "Authentication failed: ${authResult[1]}")
                        return false
                    }
                    
                    Log.d(TAG, "Authentication successful")
                }
                0xFF -> {
                    Log.e(TAG, "No acceptable authentication methods")
                    return false
                }
                else -> {
                    Log.e(TAG, "Unknown auth method: ${authResponse[1]}")
                    return false
                }
            }
            
            // Step 3: Send CONNECT request
            val connectRequest = mutableListOf<Byte>()
            connectRequest.add(0x05) // SOCKS version
            connectRequest.add(0x01) // CONNECT command
            connectRequest.add(0x00) // Reserved
            
            // Address type and address
            if (targetHost.matches(Regex("\\d+\\.\\d+\\.\\d+\\.\\d+"))) {
                // IPv4
                connectRequest.add(0x01)
                targetHost.split(".").forEach {
                    connectRequest.add((it.toInt() and 0xFF).toByte())
                }
            } else {
                // Domain name
                connectRequest.add(0x03)
                connectRequest.add(targetHost.length.toByte())
                connectRequest.addAll(targetHost.toByteArray().toList())
            }
            
            // Port
            connectRequest.add((targetPort shr 8).toByte())
            connectRequest.add((targetPort and 0xFF).toByte())
            
            output.write(connectRequest.toByteArray())
            output.flush()
            
            // Read CONNECT response
            val connectResponse = ByteArray(4)
            if (input.read(connectResponse) != 4) {
                Log.e(TAG, "Failed to read CONNECT response")
                return false
            }
            
            if (connectResponse[0] != 0x05.toByte()) {
                Log.e(TAG, "Invalid SOCKS version in response: ${connectResponse[0]}")
                return false
            }
            
            if (connectResponse[1] != 0x00.toByte()) {
                val errorMsg = when (connectResponse[1].toInt() and 0xFF) {
                    0x01 -> "General SOCKS server failure"
                    0x02 -> "Connection not allowed by ruleset"
                    0x03 -> "Network unreachable"
                    0x04 -> "Host unreachable"
                    0x05 -> "Connection refused"
                    0x06 -> "TTL expired"
                    0x07 -> "Command not supported"
                    0x08 -> "Address type not supported"
                    else -> "Unknown error: ${connectResponse[1]}"
                }
                Log.e(TAG, "SOCKS5 CONNECT failed: $errorMsg")
                return false
            }
            
            // Skip bind address (read and discard)
            val addrType = connectResponse[3].toInt() and 0xFF
            when (addrType) {
                0x01 -> { // IPv4
                    val addr = ByteArray(4)
                    input.read(addr)
                }
                0x03 -> { // Domain
                    val len = input.read()
                    if (len > 0) {
                        val domain = ByteArray(len)
                        input.read(domain)
                    }
                }
                0x04 -> { // IPv6
                    val addr = ByteArray(16)
                    input.read(addr)
                }
            }
            // Read port (2 bytes)
            val port = ByteArray(2)
            input.read(port)
            
            // Reset timeout for data transfer
            socket.soTimeout = 0 // No timeout for established connections
            
            Log.d(TAG, "✅ SOCKS5 tunnel established to $targetHost:$targetPort")
            return true
            
        } catch (e: Exception) {
            Log.e(TAG, "SOCKS5 connection error: ${e.message}", e)
            return false
        }
    }
}

interface ProxyHandler {
    fun connect(targetHost: String, targetPort: Int, socket: Socket): Boolean
}
